socorro/cron/jobs/first_signatures.py

```diff
@@ -0,0 +1,204 @@
+import datetime
+import sys
+
+import mock
+
+from configman import Namespace
+from configman.converters import class_converter
+from crontabber.base import BaseCronApp
+from crontabber.mixins import (
+    as_backfill_cron_app,
+    with_postgres_transactions
+)
+
+from socorro.external.postgresql.signature_first_date import SignatureFirstDate
+from socorro.external.es.base import ElasticsearchConfig
+from socorro.external.es.supersearch import SuperSearch
+from socorro.external.es.super_search_fields import SuperSearchFields
+from socorro.app.socorro_app import App, main
+from socorro.lib.datetimeutil import string_to_datetime
+
+
+class SuperSearchErrors(Exception):
+    """Happens when we make a SuperSearch query and there's something in the
+    results 'errors' dict."""
+
+
+@with_postgres_transactions()
+@as_backfill_cron_app
+class FirstSignaturesCronApp(BaseCronApp):
+    app_name = 'first-signatures'
+    app_description = 'Using SuperSearch to log when signatures first appeared'
+    app_version = '0.1'
+
+    required_config = Namespace()
+    required_config.add_option(
+        'window_seconds',
+        default=60,  # seconds
+        doc='Number of seconds to search for signature facets back'
+    )
+    required_config.add_option(
+        'sensitive_to_errors',
+        # default=True,
+        default=False,
+        doc=(
+            'If there are, for example, missing indexes in ES you can get errors '
+            "in supersearch doing queries across dates we don't have indices for. "
+            'If this is the case, halt this crontabber app.'
+        )
+    )
+    required_config.add_option(
+        'historic_days',
+        # The number 26 comes from the default in index_cleaner which is used
+        # as a retention policy for cleaning elasticsearch.
+        default=26 * 7,  # days
+        doc=(
+            'When looking for the oldest crash possible (by a specific signature) '
+            'go back this many days'
+        )
+    )
+    required_config.add_option(
+        'elasticsearch_class',
+        default=ElasticsearchConfig,
+    )
+
+    def run(self, date):
+        # First get a list of all unique signatures, in the given timespan
+        all_fields = SuperSearchFields(config=self.config).get()
+        api = SuperSearch(config=self.config)
+        assert isinstance(date, datetime.datetime), type(date)
+        start_date = date - datetime.timedelta(seconds=self.config.window_seconds)
+        params = {
+            'date': [
+                '>={}'.format(start_date.isoformat()),
+                '<{}'.format(date.isoformat()),
+            ],
+            '_facets': 'signature',
+            '_facets_size': 1000,
```

Comment:
This is unsatisfying. You risk missing some signatures if there are more new, different signatures in an hour's time. 

One possibility is to add another query before that to estimate the number of distinct signatures. You can do that using ``_facets=_cardinality.signature`` ([doc](https://crash-stats.mozilla.com/documentation/supersearch/#cardinality)). Note that the number it returns is an estimate, it might be safe to add some extra integers to it to make sure you get everything. 

The other possibility is to simply use an arbitrarily large number, because you want to receive all of them signatures anyway.

Reply:
So if I set _facets_size=1000 does that just limit the number of facets based on the number of crashes it matches? So to get more unique signatures, should I use `_facets=_cardinality.signature` instead and set a high `_results_number`?

Reply:
If you have a set of 11 crash results, each with a different signature, and you set ``_facet_size=10`` then you are missing out on one of those signatures. The same goes if you have 1001 distinct signatures and set size to 1000. This is the risk I'm seeing here, especially since it seems that over the last hour, we [processed more than 4000 different signatures](https://crash-stats.mozilla.com/api/SuperSearch/?_facets=_cardinality.signature&_results_number=0&date=%3E%3D2017-09-14T20%3A00%3A00&date=%3C%3D2017-09-14T21%3A00%3A00).

The number of crash reports returned by your search is unrelated to the size of the facet. You can set ``_results_number=0`` and ``_facets_size=1000`` and get your 1000 facet results.

Reply:
I'm losing confidence in this. I just now tried changing it from 1,000 to 10,000 and sure enough it found 5,282 unique signatures. Setting it to 10,000 seems to work. 

However, I don't how to use the cardinality. I tried changing `'_facets': 'signature'` to `'_facets.cardinality': 'signature'` but it was no different. 

So is it sane to use 10,000 for `_facets_size`??
I just makes me nervous when the queries are so big. Like there's something fundamentally wrong. 

Other crazy ideas? What about instead of running this crontabber app every hour, we run it every 30 minutes. Or 10 minutes. With Prod ES and a 30 min window I get 3,101 unique signatures. For 10 minutes, 1,330 unique signatures.

Reply:
I don't have a good answer for you here. Setting an arbitrarily high number for ``_facets_size`` is a good thing to do, to ensure you don't miss anything. Running it more regularly is probably good too, though I suspect, as your numbers show, that we won't get much less results, since many signatures overlap. 

To use ``_cardinality.signature`` you would have to make another query, in order to get only that number. It's probably pointless, since we want to get all the results anyway. Using an arbitrary number is probably better.